README.md

```diff
@@ -1,8 +1,380 @@
+
+The builtin components of Zanzibar makes it easy to develop mircoserivces and gateway services that proxy or orchestrate microsevices. It is also simple to extend Zanzibar with custom plugins to ease the development of applications that suit your specific needs.
+
+## Concepts
+Zanzibar is built on three pillars: module, config, code generation.
+### Module
+Modules are the components that a Zanzibar application is made of. A module belongs to a `ModuleClass`, has a `type` and can have dependencies on other modules.
+
+#### ModuleClass
+`ModuleClass` abstracts the functionality of a specific class of components. Zanzibar predefines a few module classes, i.e., `client`, `endpoint`, `middleware` and `service`. Each represents a corresponding abstraction:
+
+ModuleClass | Abstraction
+----------- | -----------
+client | clients to communicate with downstreams, e.g., database clients and RPC clients
+endpoint | application interfaces exposed to upstreams
+middleware | common functionality that has less to do with business logic, e.g., rate limiting middleware
+servcie | high level appliation abstraction, e.g., a demo service that prints "Hello World!"
+
+#### Type
+The module `type` differentiates module instances of the same `ModuleClass` with further classification.
+
+For example, a client module could be of type `http`, `tchannel` or `custom`, where `http` or `tchannel` means Zanziabr will generate a client with given configuration that speaks that protocol while `custom` means the client is fully provided and Zanzibar will use it as is without code generation.
+
+An `endpoint` module could also be of type `http` or `tchannel`, which represents the endpoint's runtime protocol. While `endpoint` modules do not have `custom` type, each method of an `endpoint` module has a `workflowType` that indicates the type of workflow the endpoint method fulfills. The builtin workflow type is `httpClient`, `tchannelClient` and `custom`, where `httpClient` and `tchannelClient` means the endpoint method workflow is to proxy to a client, and `custom` means the workflow is fulfilled by user code, see more in [Custom Workflow](#custom-workflow). **Note** that workflow type is likely to be deprecated in the future so that proxy to a client will be no longer a builtin option.
+
+The builtin type of middleware module is `default` and the builtin service type is `gateway` (`default` is probably a better name, up to change in the future). As you maybe already noticed, the types are somewhat arbitrary as they are not necessarily abstractions but indications about how Zanzibar should treat the modules.
+
+Note that the module classes and types mentioned above are builtin in the sense that Zanzibar knows how to treat a module with given builtin classes and types. We can always easily extend by defining arbitrary module classs and types as long as we register appropriate module generators to the module system.
```

Comment:
you should rephrase as:

> Zanzibar has support for user-defined module classes and module types in case the builtin types are not sufficient. The preferred way of extending Zanzibar is through user defined module classes and module types.

Reply:
done